connmgr/conngating.go

```diff
@@ -0,0 +1,43 @@
+package connmgr
+
+import (
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// AddressConnectionGater can be implemented by a type that supports active
+// connection gating based on the multi-address of the peer.
+type AddressConnectionGater interface {
+	// DenyAddrConnection returns true if a connection to/from a peer
+	// should be denied.
+	// The caller must supply the multi-address of the remote peer.
+	DenyAddrConnection(ma.Multiaddr) (deny bool)
+}
+
+// PeerConnectionGater can be implemented by a type that supports active
+// inbound or outbound connection gating based on the direction of the connection and the Identity
+// of the remote peer.
+type PeerConnectionGater interface {
+	// DenyPeerConnection returns true if a connection to/from a peer
+	// should be denied.
+	// The caller must supply the direction of the connection and the Id of the remote peer.
+	DenyPeerConnection(direction network.Direction, id peer.ID) (deny bool)
+}
+
+// ConnectionGater can be implemented by a type that supports active
+// inbound or outbound connection gating.
+//
+// A ConnectionGater can be consulted by the
+// network.Network (a) before initiating a connection attempt for outbound
+// connections, or (b) incorporating an inbound connection into the node's
+// state.
+//
+// This feature can be used to implement *strict* connection management
+// behaviours, such as hard limiting of connections once a max count has been
+// reached.
+type ConnectionGater interface {
```

Comment:
@raulk Once we have the Disconnect message in the Control protocol, I'm inclined to add a `Disconnectpb.Reason` field to the return values of the Gaters so that the caller can send a Disconnect message with the appropriate reason if the connection gets gated.

Is the idea stupid ?